File: frontend/src/services/ingestions.ts
/**
 * 导入任务 API 封装
 */

import { apiFetch } from './http';

export type IngestionStatus = 'queued' | 'running' | 'succeeded' | 'failed' | 'canceled';
export type IngestionMode = 'create' | 'update';

export interface IngestionJob {
  id: string;
  kb_id: string;
  status: IngestionStatus;
  error_message: string | null;
  stats: Record<string, unknown> | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

export interface IngestionJobCreateRequest {
  kb_id: string;
  material_ids: string[];
  mode?: IngestionMode;
}

/**
 * 创建导入任务
 */
export async function createIngestionJob(
  data: IngestionJobCreateRequest
): Promise<IngestionJob> {
  return apiFetch<IngestionJob>('/api/v1/ingestions', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

/**
 * 获取导入任务状态
 */
export async function getIngestionJob(ingestionId: string): Promise<IngestionJob> {
  return apiFetch<IngestionJob>(`/api/v1/ingestions/${ingestionId}`);
}

/**
 * 取消导入任务
 */
export async function cancelIngestionJob(ingestionId: string): Promise<IngestionJob> {
  return apiFetch<IngestionJob>(`/api/v1/ingestions/${ingestionId}/cancel`, {
    method: 'POST',
  });
}

/**
 * 轮询导入任务状态直到完成
 */
export async function pollIngestionJob(
  ingestionId: string,
  options?: { intervalMs?: number; maxAttempts?: number }
): Promise<IngestionJob> {
  const { intervalMs = 2000, maxAttempts = 60 } = options ?? {};
  let attempts = 0;

  while (attempts < maxAttempts) {
    const job = await getIngestionJob(ingestionId);
    if (['succeeded', 'failed', 'canceled'].includes(job.status)) {
      return job;
    }
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
    attempts++;
  }

  throw new Error('导入任务轮询超时');
}

File: backend/src/app/schemas/ingestions.py
"""导入任务相关 Schemas。"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.pagination import PageMeta


class IngestionStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELED = "canceled"


class IngestionMode(str, Enum):
    CREATE = "create"
    UPDATE = "update"


class IngestionJobCreateRequest(BaseModel):
    """创建导入任务请求。"""

    kb_id: uuid.UUID
    material_ids: list[uuid.UUID] = Field(..., min_length=1)
    mode: IngestionMode = IngestionMode.CREATE


class IngestionJobRead(BaseModel):
    """导入任务读取响应。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    kb_id: uuid.UUID
    status: IngestionStatus
    error_message: str | None = None
    stats: dict | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None


class IngestionJobListResponse(BaseModel):
    """导入任务列表响应。"""

    items: list[IngestionJobRead]
    page: PageMeta
